pip/commands/install.py

```diff
@@ -88,8 +88,20 @@ def __init__(self, *args, **kw):
             dest='upgrade',
             action='store_true',
             help='Upgrade all specified packages to the newest available '
-                 'version. This process is recursive regardless of whether '
-                 'a dependency is already satisfied.'
+                 'version. The handling of dependencies depends on the '
+                 'upgrade-strategy used.'
+        )
+
+        cmd_opts.add_option(
+            '--upgrade-strategy',
+            dest='upgrade_strategy',
+            default='non-eager',
```

Comment:
As I mentioned in https://github.com/pypa/pip/issues/3871#issuecomment-247789343 I think we'll want this to default to ``eager`` for this time and we'll make sure to file an issue to swap the default for the next X+1 release.